### PR TITLE
Akcore 60 : Add unit tests for acknowledge function in Share Partition Manager

### PR DESCRIPTION
**Pull request by chirag-wadhwa5**
This PR contains unit tests for acknowledge function in SharePartitionMaager

Reference: [AKCORE-60](https://confluentinc.atlassian.net/jira/software/c/projects/AKCORE/issues/AKCORE-60)

[AKCORE-60]: https://confluentinc.atlassian.net/browse/AKCORE-60?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: core/src/main/java/kafka/server/SharePartitionManager.java [hunk 1]
```diff
@@ -68,11 +67,11 @@ public class SharePartitionManager {
     private final ConcurrentLinkedQueue<ShareFetchPartitionData> fetchQueue;
     private final AtomicBoolean processFetchQueueLock;
 
-    public SharePartitionManager(ReplicaManager replicaManager, Time time, ShareSessionCache cache) {
+    public SharePartitionManager(ReplicaManager replicaManager, Time time, Map<SharePartitionKey, SharePartition> partitionCacheMap, ShareSessionCache cache) {
```
Comment (adixitconfluent) [inline core/src/main/java/kafka/server/SharePartitionManager.java:70]: @apoorvmittal10 do you think it makes sense that we add `partitionCacheMap` to the constructor of SharePartitionManager, when in reality, it has to be populated by fetch/acknowledge operations within the SharePartitionManager. I see it in the constructor call in `BrokerServer.scala`, it is always initialized as an empty map

**Review by adixitconfluent — CHANGES_REQUESTED** [review]

File: core/src/main/java/kafka/server/SharePartitionManager.java (hunk 1 above)
Comment (adixitconfluent) [reply core/src/main/java/kafka/server/SharePartitionManager.java:70]: Maybe just making `partitionCacheMap` as Default might be better?

Comment (apoorvmittal10) [reply core/src/main/java/kafka/server/SharePartitionManager.java:70]: The way to do this this adding an overloaded constructor which takes:
1. `SharePartitionManager(ReplicaManager replicaManager, Time time, Map<SharePartitionKey, SharePartition> partitionCacheMap, ShareSessionCache cache)` arguments. 
2. The old public constructor should not expose `Map<SharePartitionKey, SharePartition> partitionCacheMap` hence no need to change anywhere else in the code.
3. The new overloaded constructor should be of `default` visibility scope so it's exposed to tests.
4. The new overloaded constructor should have a comment on top `// Visible for testing`. You will see such instances in the current code as well.

File: core/src/main/scala/kafka/server/BrokerServer.scala [hunk 2]
```diff
@@ -394,7 +394,8 @@ class BrokerServer(
 
       val shareFetchSessionCache : ShareSessionCache = new ShareSessionCache(config.shareGroupMaxGroups * config.shareGroupMaxSize,
         KafkaServer.MIN_INCREMENTAL_FETCH_SESSION_EVICTION_MS)
-      val sharePartitionManager = new SharePartitionManager(replicaManager, Time.SYSTEM, shareFetchSessionCache);
+      val partitionCacheMap: ConcurrentHashMap[SharePartitionKey, SharePartition] = new ConcurrentHashMap[SharePartitionKey, SharePartition]()
+      val sharePartitionManager = new SharePartitionManager(replicaManager, Time.SYSTEM, partitionCacheMap, shareFetchSessionCache)
```
Comment (apoorvmittal10) [inline core/src/main/scala/kafka/server/BrokerServer.scala:398]: No changes should be required here.

File: core/src/test/java/kafka/server/SharePartitionManagerTest.java [hunk 3]
```diff
@@ -1128,4 +1135,53 @@ public void testShareFetchContextResponseSize() {
         assertEquals(Errors.NONE, resp8.error());
         assertEquals(4 + resp8.data().size(objectSerializationCache, version), respSize8);
     }
+
+    @Test
+    public void testAcknowledge() {
```
Comment (apoorvmittal10) [inline core/src/test/java/kafka/server/SharePartitionManagerTest.java:1140]: I am sure there would be tons of unit tests for `Acknowledge`, are we planning to add further in next PR?

**Review by apoorvmittal10 — CHANGES_REQUESTED** [review]

File: core/src/main/java/kafka/server/SharePartitionManager.java [hunk 4]
```diff
@@ -74,6 +74,16 @@ public SharePartitionManager(ReplicaManager replicaManager, Time time, ShareSess
         this.time = time;
         this.cache = cache;
         this.partitionCacheMap = new ConcurrentHashMap<>();
+        fetchQueue = new ConcurrentLinkedQueue<>();
```
Comment (apoorvmittal10) [inline core/src/main/java/kafka/server/SharePartitionManager.java:77]: Please maintain consistency `this.`.

File: core/src/main/java/kafka/server/SharePartitionManager.java [hunk 5]
```diff
@@ -74,6 +74,16 @@ public SharePartitionManager(ReplicaManager replicaManager, Time time, ShareSess
...
+        this.processFetchQueueLock = new AtomicBoolean(false);
+    }
```
Comment (apoorvmittal10) [inline core/src/main/java/kafka/server/SharePartitionManager.java:79]: Also why not to call the other constructor from public constructor by calling `this(replicaManager, time, cache, new ConcurrentHashMap<>());`

**Review by apoorvmittal10 — CHANGES_REQUESTED** [review]

**Review by apoorvmittal10 — APPROVED** [review]
Lgtm

**Pull request merged**